[8.1] refactor: replacing voimport from Utils by ObjectLoader class

simon-mazenoux opened the pull request:
Small refactor, no need for a release note in my opinion.

The function voimport was using the __import__ statement, but the python doc suggest that the programmatic importing of modules should use the import_module instead of __import__. (https://docs.python.org/3/library/importlib.html#importlib.__import__). The doc specifies that there are minor differences between the two functions, but the change didn't affect my tests. Therefore, I think it should be best that we switch the import method.

chrisburr commented:
Can you also take a look at switching the other occurances of `__import__` to use `importlib.import_module` instead?

simon-mazenoux commented:
> 
When looking at the import statements, i notice that a lot of code doesn't use the proper methods to import a class object in dirac. 
Also, i noticed that there is two ways in dirac to import a object:
- voimport (in src/DIRAC/ResourceStatusSystem/Utilities/Utils.py)
- ObjectLoader class (at src/DIRAC/Core/Utilities/ObjectLoader.py)

I suppose we should keep only one of these two, and I suppose that we should keep ObjectLoader. What do you think about this issue ?

chrisburr commented:
Yeah I agree that `voimport` can be removed and that this can all be unified 👍

fstagni commented:
Yes! we actually had a long standing issue on this: https://github.com/DIRACGrid/DIRAC/issues/3514

simon-mazenoux commented:
Looking a little bit more at the code, i found another ObjectLoader type file named ModuleFactory (/home/simon/CERN/DIRAC/src/DIRAC/Core/Utilities/ModuleFactory.py). Should I create a new branch or should i keep working on this branch ?

fstagni commented:
> Looking a little bit more at the code, i found another ObjectLoader type file named ModuleFactory (/home/simon/CERN/DIRAC/src/DIRAC/Core/Utilities/ModuleFactory.py). Should I create a new branch or should i keep working on this branch ?

You can do it in this same PR

simon-mazenoux commented:
I don't understand how the ModuleFactory.getModule() (src/DIRAC/Core/Utilities/ModuleFactory.py) works. The argumentDict parameter is unused, the docstring says the modules are instanciated and all the functions use the method .execute() which I can't find anywhere in the python doc. Could someone give me a hint about how this code works ?

fstagni commented:
> I don't understand how the ModuleFactory.getModule() (src/DIRAC/Core/Utilities/ModuleFactory.py) works. The argumentDict parameter is unused, the docstring says the modules are instanciated and all the functions use the method .execute() which I can't find anywhere in the python doc. Could someone give me a hint about how this code works ?

This is probably a very old module. I see it uses https://docs.python.org/3/library/functions.html#eval and this is where the `argumentDict` is consumed. Not the best way of doing things.

chrisburr commented:
> The argumentDict parameter is unused

The `argumentDict` is being used by the `eval` (which should never have been used in DIRAC).

> the docstring says the modules are instanciated and all the functions use the method .execute() which I can't find anywhere in the python doc. Could someone give me a hint about how this code works ?

This module is using the convention that DIRAC has that modules with the same name as the class they contain (e.g. `DIRAC.WorkloadManagementSystem.Client.InputDataResolution.InputDataResolution`).

If I ignore the error handling there are the 5 main lines:

```python
# Split the string to get the parent import (modulePath) and the name of the class in that module (moduleName)
moduleName = importString.split(".")[-1]
modulePath = importString.replace(".%s" % (moduleName), "")
# Do the actual import of the module
importModule = __import__("%s.%s" % (modulePath, moduleName), globals(), locals(), [moduleName])
# Get the class and call initialise it with argumentsDict
moduleStr = "importModule.%s(argumentsDict)" % (moduleName)
moduleInstance = eval(moduleStr)
```

I haven't tested it, but this should be equivilent:

```python
moduleName = importString.split(".")[-1]
importModule = importlib.import_module(importString)
moduleInstance = getattr(importModule, moduleName)(argumentsDict)
```

simon-mazenoux commented:
> > The argumentDict parameter is unused
> 
> The `argumentDict` is being used by the `eval` (which should never have been used in DIRAC).
> 
> > the docstring says the modules are instanciated and all the functions use the method .execute() which I can't find anywhere in the python doc. Could someone give me a hint about how this code works ?
> 
> This module is using the convention that DIRAC has that modules with the same name as the class they contain (e.g. `DIRAC.WorkloadManagementSystem.Client.InputDataResolution.InputDataResolution`).
> 
> If I ignore the error handling there are the 5 main lines:
> 
> ```python
> # Split the string to get the parent import (modulePath) and the name of the class in that module (moduleName)
> moduleName = importString.split(".")[-1]
> modulePath = importString.replace(".%s" % (moduleName), "")
> # Do the actual import of the module
> importModule = __import__("%s.%s" % (modulePath, moduleName), globals(), locals(), [moduleName])
> # Get the class and call initialise it with argumentsDict
> moduleStr = "importModule.%s(argumentsDict)" % (moduleName)
> moduleInstance = eval(moduleStr)
> ```
> 
> I haven't tested it, but this should be equivilent:
> 
> ```python
> moduleName = importString.split(".")[-1]
> importModule = importlib.import_module(importString)
> moduleInstance = getattr(importModule, moduleName)(argumentsDict)
> ```

Thanks for this very thorough explanation :-). However, I still have a question: should I use the ObjectLoader class or just use theses three lines of python code ? Personally, I think using ObjectLoader is best, but is it possible that using ObjectLoader could bring issues with DIRAC extensions ?

chrisburr commented:
> Thanks for this very thorough explanation :-). However, I still have a question: should I use the ObjectLoader class or just use theses three lines of python code ? Personally, I think using ObjectLoader is best, but is it possible that using ObjectLoader could bring issues with DIRAC extensions ?

I think the `ObjectLoader` shouldn't be used in `ModuleFactory` as it currently doesn't replace classes with extensions and this appears to be intentional.

fstagni commented on src/DIRAC/ResourceStatusSystem/PolicySystem/Actions/LogPolicyResultAction.py:
```diff
@@ -18,14 +14,16 @@ class LogPolicyResultAction(BaseAction):
+            result = ObjectLoader().loadObject("DIRAC.ResourceStatusSystem.Client.ResourceManagementClient")
+            if not result["OK"]:
+                raise Exception(result["Value"])
+            ressourceManagementClientClass = result["Value"]
+            self.rmClient = ressourceManagementClientClass()
```
```suggestion
            resourceManagementClientClass = result["Value"]
            self.rmClient = resourceManagementClientClass()
```

fstagni commented on src/DIRAC/ResourceStatusSystem/Service/PublisherHandler.py:
```diff
@@ -1,37 +1,37 @@
     @classmethod
     def initializeHandler(cls, serviceInfoDict):
         """Initialization of client objects"""
         cls.rsClient = ResourceStatusClient()
```
Also this could be loaded via ObjectLoader.

simon-mazenoux replied:
Concerning this class, should I raise an exception or should I return S_ERROR ? It's not an `__init__` method, and it already returns S_OK but I didn't manage to find where this function is called.

fstagni replied:
You should return `S_ERROR`.

fstagni commented on src/DIRAC/AccountingSystem/Client/AccountingCLI.py:
```diff
@@ -71,12 +72,10 @@ def do_registerType(self, args):
-                typeModule = __import__(
-                    "DIRAC.AccountingSystem.Client.Types.%s" % typeName, globals(), locals(), typeName
-                )
+                typeModule = importlib.import_module(f"DIRAC.AccountingSystem.Client.Types.{typeName}")
```
I think that in this module we should use everywhere ObjectLoader.

simon-mazenoux commented on src/DIRAC/ResourceStatusSystem/Command/CommandCaller.py:
```diff
@@ -30,20 +30,17 @@ def commandInvocation(commandTuple, pArgs=None, decisionParams=None, clients=Non
+    command = commandAttribute(newArgs, clients)
 
-    return S_OK(commandObject)
+    return S_OK(command)
```
Yes it is, twice actually

simon-mazenoux commented on src/DIRAC/ResourceStatusSystem/DB/ResourceManagementDB.py:
```diff
@@ -402,11 +401,16 @@ def __init__(self, parentLogger=None):
-        self.tablesList = getattr(Utils.voimport("DIRAC.ResourceStatusSystem.DB.ResourceManagementDB"), "TABLESLIST")
+        result = self.objectLoader.loadObject("DIRAC.ResourceStatusSystem.DB.ResourceManagementDB", "TABLESLIST")
+        if not result["OK"]:
+            raise Exception(result["Value"])
```
The loadObject method can either raise an AttributeErrror or an ImportError, so I don't really know how we could specialize here if we still want the loadObject method to return S_OK / S_ERROR.

fstagni replied:
This would be `AttributeError`, isn't it?

fstagni commented on src/DIRAC/ResourceStatusSystem/Service/PublisherHandler.py:
```diff
@@ -1,37 +1,43 @@
+
+        result = objectLoader.loadObject("DIRAC.ResourceStatusSystem.Client.ResourceStatusClient")
+        if not result["OK"]:
+            raise Exception(result["Value"])
```
```suggestion
            raise Exception(result["Message"])
```

fstagni commented on src/DIRAC/ResourceStatusSystem/Service/PublisherHandler.py:
```diff
@@ -1,37 +1,43 @@
+
+        result = objectLoader.loadObject("DIRAC.ResourceStatusSystem.Client.ResourceManagementClient")
+        if not result["OK"]:
+            raise Exception(result["Value"])
```
```suggestion
            raise Exception(result["Message"])
```

fstagni commented on src/DIRAC/TransformationSystem/Client/WorkflowTasks.py:
```diff
@@ -460,31 +465,38 @@ def _handleHospital(self, oJob, clinicPath):
         if not self.outputDataModule_o:
-            # Create the module object
-            moduleFactory = ModuleFactory()
+            # TODO: check with Federico if we shouldn't use ObjectLoader every time
```
It would make sense to use `ObjectLoader`

simon-mazenoux replied:
Should I also use ObjectLoader for the `__generatePluginObject` method or not ?

fstagni replied:
yes

fstagni commented:
There are some conflicts now, please rebase.

fstagni commented on src/DIRAC/Resources/Computing/BatchSystems/TimeLeft/test/Test_TimeLeft.py:
```diff
@@ -2,12 +2,15 @@
+
+# pylint: disable=invalid-name
+
 from __future__ import absolute_import
 from __future__ import division
 from __future__ import print_function
```
You can remove these imports.

fstagni commented on src/DIRAC/TransformationSystem/Agent/TransformationAgent.py:
```diff
@@ -209,10 +210,10 @@ def _execute(self, transDict):
         # Each thread will have its own clients
         clients = self._getClients()
 
+        startTime = time.time()
```
Don't move this, if not necessary.

fstagni commented on src/DIRAC/TransformationSystem/Agent/TransformationAgent.py:
```diff
@@ -699,22 +698,24 @@ def __writeCache(self, transID=None):
-        plugin_o.setCallback(self.pluginCallback)
+
+        plugin_o = pluginClass(
+            "%s" % plugin, transClient=clients["TransformationClient"], dataManager=clients["DataManager"]
```
```suggestion
            plugin, transClient=clients["TransformationClient"], dataManager=clients["DataManager"]
```

fstagni commented:
BTW there's a conflict now, please rebase.

chrisburr commented on src/DIRAC/Core/Base/Client.py:
```diff
@@ -3,8 +3,19 @@
+
+try:
+    from functools import partialmethod
+except ImportError:
+
+    class partialmethod(partial):
+        def __get__(self, instance, owner):
+            if instance is None:
+                return self
+            return partial(self.func, instance, *(self.args or ()), **(self.keywords or {}))
```
I think this is from a bad rebase as this was a Python 2 compaibility hack that was removed from 8.x in ce39d7e805f40e5fc28407f7fa1573a44661efff.
